dmpbbo/dmps/DmpWithSchedules.py

```diff
@@ -0,0 +1,280 @@
+        schedules = np.ndarray((n_schedules,))
+        for i_dim in range(n_schedules):
+            schedules[i_dim] = self._func_apps_schedules[i_dim].predict(xs[self.PHASE])
+        schedules = self._enforce_schedule_bounds(schedules)
+
+        return xs, xds, schedules
+
+    def integrate_step_sched(self, dt, x):
+        """ Integrate the system one time step.
+
+        @param dt: Duration of the time step
+        @param x: Current state
+        @return: x_updated, xd_updated, schedules - Updated state and its rate of change,
+        as well as the schedules, dt time later
+        """
+        x, xd = super().integrate_step(dt, x)
+
+        schedules = np.ndarray((len(self._func_apps_schedules),))
+        for i_dim in range(len(self._func_apps_schedules)):
+            schedules[i_dim] = self._func_apps_schedules[i_dim].predict(x[self.PHASE])
+        schedules = self._enforce_schedule_bounds(schedules)
+
+        return x, xd, schedules
+
+    def states_as_trajectory_sched(self, ts, x_in, xd_in, schedules):
+        """Get the output of a DMP dynamical system as a trajectory, including schedules.
+
+        As it is a dynamical system, the state vector of a DMP contains the output of the goal,
+        spring, phase and gating system. What we are most interested in is the output of the
+        spring system. This function extracts that information, and also computes the
+        accelerations of the spring system, which are only stored implicitly in xd_in because
+        second order systems are converted to first order systems with expanded state.
+
+        The schedules are added as misc variables to the trajectory.
+
+        @param ts: A vector of times
+        @param x_in: State vector over time
+        @param xd_in: State vector over time (rates of change)
+        @param schedules: Gain/force schedules over time
+
+        Return:
+            Trajectory representation of the DMP state vector output.
+        """
+        traj = super().states_as_trajectory(ts, x_in, xd_in)
+        traj.misc = schedules
+        return traj
+
+    def set_selected_param_names(self, names):
+        """ Get the names of the parameters that have been selected.
+
+        @param names: Names of the parameters that have been selected.
+        """
+        if isinstance(names, str):
+            names = [names]  # Convert to list
+
+        # Pass all names starting with "sched_" to the function approximators
+        # (but without the prefix "sched_")
+        sched_names = [n[len("sched_") :] for n in names if "sched_" in n]
+        for fa in self._func_apps_schedules:
+            fa.set_selected_param_names(sched_names)
+
+        # Pass all names not starting with "sched_" to the Dmp superclass
+        other_names = [n for n in names if "sched_" not in n]
+        super().set_selected_param_names(other_names)
+
+    def get_param_vector(self):
+        """Get a vector containing the values of the selected parameters."""
```

Comment:
If calling set_selected_param_names is needed before get_param_vector, I would write more clear here

Reply:
Converted this into #75, as it requires minor refactoring of all classes that inherit from Parameterizable (not just the ones in this PR)